buffer/src/main/java/io/netty/buffer/AbstractLeakAwareByteBuf.java

```diff
@@ -0,0 +1,157 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer;
+
+import io.netty.util.ResourceLeakTracker;
+import io.netty.util.internal.ObjectUtil;
+
+import java.nio.ByteOrder;
+
+abstract class AbstractLeakAwareByteBuf extends WrappedByteBuf {
+
+    final ResourceLeakTracker<ByteBuf> leak;
+
+    AbstractLeakAwareByteBuf(ByteBuf buf, ResourceLeakTracker<ByteBuf> leak) {
+        super(buf);
+        this.leak = ObjectUtil.checkNotNull(leak, "leak");
+    }
+
+    @Override
+    public ByteBuf slice() {
+        return newLeakAwareByteBuf(super.slice(), leak);
+    }
+
+    @Override
+    public ByteBuf retainedSlice() {
+        return unwrappedDerived(super.retainedSlice());
+    }
+
+    @Override
+    public ByteBuf retainedSlice(int index, int length) {
+        return unwrappedDerived(super.retainedSlice(index, length));
+    }
+
+    @Override
+    public ByteBuf retainedDuplicate() {
+        return unwrappedDerived(super.retainedDuplicate());
+    }
+
+    @Override
+    public ByteBuf readRetainedSlice(int length) {
+        return unwrappedDerived(super.readRetainedSlice(length));
+    }
+
+    @Override
+    public ByteBuf slice(int index, int length) {
+        return newLeakAwareByteBuf(super.slice(index, length), leak);
+    }
+
+    @Override
+    public ByteBuf duplicate() {
+        return newLeakAwareByteBuf(super.duplicate(), leak);
+    }
+
+    @Override
+    public ByteBuf readSlice(int length) {
+        return newLeakAwareByteBuf(super.readSlice(length), leak);
+    }
+
+    @Override
+    public ByteBuf asReadOnly() {
+        return newLeakAwareByteBuf(super.asReadOnly(), leak);
+    }
+
+    @Override
+    public ByteBuf touch() {
+        return this;
+    }
+
+    @Override
+    public ByteBuf touch(Object hint) {
+        return this;
+    }
+
+    @Override
+    public boolean release() {
+        // Call unwrap() before just in case that super.release() will change the ByteBuf instance that is returned
+        // by ByteBuf.
+        ByteBuf unwrapped = unwrap();
+        if (super.release()) {
+            boolean closed = leak.close(unwrapped);
+            assert closed;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean release(int decrement) {
+        // Call unwrap() before just in case that super.release() will change the ByteBuf instance that is returned
+        // by ByteBuf.
+        ByteBuf unwrapped = unwrap();
+        if (super.release(decrement)) {
+            boolean closed = leak.close(unwrapped);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "closed". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)